spec.bs

```diff
@@ -317,18 +323,22 @@ This method creates a {{HandwritingRecognizer}} object that satisfies the provid
 <div algorithm="navigator-create-handwriting-recognizer">
   When {{Navigator/createHandwritingRecognizer(constraint)}} method is invoked, do the following:
 
-  1. If |constraint| doesn't have a {{HandwritingModelConstraint/languages}} member, [=exception/throw=] a new {{TypeError}}
+  1. If |constraint| doesn't have a {{HandwritingModelConstraint/languages}} member, return [=a promise rejected with=] {{TypeError}}.
   1. Let |p| be [=a new promise=].
-  1. Run the following step [=in parallel=]
-    1. If |constraint|'s {{HandwritingModelConstraint/languages}} is an empty array, [=/reject=] |p| with a new {{DOMException}} whose {{DOMException/name}} attribute is {{"NotSupportedError"}} and abort.
-    1. <a lt="convert">Convert |constraint| into a suitable form</a> for [=handwriting recognizer=].
-    1. If the user agent can't find a [=handwriting recognizer=] that satisfies the converted |constraint|, [=/reject=] |p| with a new {{DOMException}} whose {{DOMException/name}} attribute is {{"NotSupportedError"}} and abort.
-    1. If creating a [=handwriting recognizer=] would cause the user agent to exceed its limit for total number of active recognizer, [=/reject=] |p| with a new {{DOMException}} whose {{DOMException/name}} attribute is {{"QuotaExceededError"}}.
-    1. If the user agent can't prepare [=handwriting recognizer=] to perform recognitions, [=/reject=] |p| with a new {{DOMException}} object whose name depending on the failure cause and abort:
-      * If the web application shouldn't retry, {{"OperationError"}}.
-      * Otherwise {{"UnknownError"}}.
-    1. Let |result| be a new {{HandwritingRecognizer}} object, whose [=HandwritingRecognizer/active=] flag is `true`
-    1. [=/Resolve=] |p| with |result|.
+  1. Run the following step [=in parallel=]:
+    1. <a lt="convert">Convert |constraint| into a suitable form</a> for creating a platform-dependent [=handwriting recognizer=].
+    1. If the user agent can't create or prepare a [=handwriting recognizer=] to perform recognitions, [=queue a Handwriting Recognition API task=] to [=/reject=] |p| with a new {{DOMException}} according to the failure cause, then abort the remaining steps:
```

Comment:
Both steps "queue a Handwriting Recognition API task". Maybe worth re-writing as:

1. [=Queue a Handwriting Recognition API task=] to:
    1. If the user agent can't create or prepare a [=handwriting recognizer=] to perform recognitions, [=/reject=] |p| with a new {{DOMException}} according to the failure cause:
        * If |constraint|'s ...
    2. Otherwise:
        1. Let |result| be...

Reply:
Would this cause the sequence of event to be interpreted differently?

I think the "if checks" (e.g. check if user agent can instantiate a suitable recognizer) should happen outside of the enqueued tasks, because those checks has nothing to do with JavaScript / DOM.

Reply:
I don't think the change introduces any substantive difference, but I may have missed something ;)

The [Infra Standard](https://infra.spec.whatwg.org/#algorithm-conformance) explains that "Conformance requirements phrased as algorithms or specific steps may be implemented in any manner, so long as the end result is equivalent. (In particular, the algorithms are intended to be easy to follow, and not intended to be performant.)"

Where and how the user agent checks the if condition is not observable by the application as far as I can tell. In particular, it does not trigger any event that the application could see. Regardless of how the algorithm is written, it seems totally fine for an implementation to check that condition within the enqueued task or outside of the enqueued task.

Reply:
Ah. Thanks for the explanation. I've updated to use

```
Queue a task to:
- If, ...
- Otherwise, ...
```